File: app/src/main/java/com/klikacc/android/di/modules/NetworkModule.java
package com.klikacc.android.di.modules;

import com.google.gson.Gson;
import com.klikacc.android.BuildConfig;

import java.util.concurrent.TimeUnit;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class NetworkModule {

    private static final String BASE_ENDPOINT = "http://www.google.com";

    @Singleton
    @Provides
    Gson provideGson() {
        return new Gson();
    }

    @Singleton
    @Provides
    OkHttpClient provideOkHttpClient() {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(BuildConfig.DEBUG
                ? HttpLoggingInterceptor.Level.BODY
                : HttpLoggingInterceptor.Level.NONE);

        return new OkHttpClient.Builder()
                .connectTimeout(40, TimeUnit.SECONDS)
                .writeTimeout(40, TimeUnit.SECONDS)
                .readTimeout(40, TimeUnit.SECONDS)
                .addInterceptor(loggingInterceptor)
                .build();
    }

    @Singleton
    @Provides
    Retrofit provideRetrofitV1(OkHttpClient okHttpClient, Gson gson) {
        return getRetrofit(okHttpClient, BASE_ENDPOINT, gson);
    }

    private Retrofit getRetrofit(OkHttpClient okHttpClient, String endpoint, Gson gson) {
        return new Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(endpoint)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .build();
    }
}

File: app/src/main/java/com/esafirm/androidboilerplate/di/module/NetworkModule.kt
package com.esafirm.androidboilerplate.di.module

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.esafirm.androidboilerplate.data.remote.ApiService
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

@Module class NetworkModule {

	private val ENDPOINT = "https://api.ribot.io/"

	@Singleton @Provides internal fun provideGson(): Gson {
		return GsonBuilder()
				.setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
				.create()
	}

	@Singleton @Provides internal fun provideOkHttpClient(): OkHttpClient {
		val loggingInterceptor = HttpLoggingInterceptor { message -> Timber.d(message) }
		loggingInterceptor.level = HttpLoggingInterceptor.Level.BODY

		return OkHttpClient.Builder()
				.connectTimeout(40, TimeUnit.SECONDS)
				.writeTimeout(40, TimeUnit.SECONDS)
				.readTimeout(40, TimeUnit.SECONDS)
				.addInterceptor(loggingInterceptor)
				.build()
	}

	@Singleton @Provides internal fun provideRetrofit(gson: Gson, okHttpClient: OkHttpClient): Retrofit {
		return Retrofit.Builder()
				.client(okHttpClient)
				.baseUrl(ENDPOINT)
				.addConverterFactory(GsonConverterFactory.create(gson))
				.addCallAdapterFactory(RxJavaCallAdapterFactory.create())
				.build()
	}

	/* --------------------------------------------------- */
	/* > Retrofit Interfaces*/
	/* --------------------------------------------------- */

	@Singleton @Provides fun provideApiService(retrofit: Retrofit): ApiService {
		return retrofit.create(ApiService::class.java)
	}
}
